File: 07-bridge/07-bridge.go
package main

import "fmt"

type Notification interface {
	send(message string)
}

type Email struct{}

func (e *Email) send(message string) {
	fmt.Println("Sending email: " + message)
}

type SMS struct{}

func (s *SMS) send(message string) {
	fmt.Println("Sending SMS: " + message)
}

type Importance interface {
	sendMessage(message string)
}

type OrderNotification struct {
	Notification Notification
}

func (o *OrderNotification) sendMessage(message string) {
	o.Notification.send(message)
	fmt.Println("Order notification sent.")
}

type ShippingNotification struct {
	Notification Notification
}

func (s *ShippingNotification) sendMessage(message string) {
	s.Notification.send(message)
	fmt.Println("Shipping notification sent.")
}

func main() {
	e := &Email{}
	o := OrderNotification{Notification: e}
	o.sendMessage("Order #123 has been processed.")

	s := &SMS{}
	n := ShippingNotification{Notification: s}
	n.sendMessage("Order #236 has been shipped.")
}

File: 07-bridge/07-bridge.js
class Notification {
    constructor() {
        if (new.target === Notification) {
            throw new Error("Cannot instantiate an abstract class");
        }
    }

    send(message) {
        throw new Error("Method not implemented");
    }
}

class Email extends Notification {
    send(message) {
        return "Sending Email: " + message;
    }
}

class SMS extends Notification {
    send(message) {
        return "Sending SMS: " + message;
    }
}

// Abstraction: Urgency levels
class Urgency {
    constructor(notification) {
        this.notification = notification;
    }

    sendMessage(message) {
        throw new Error("not implemented");
    }
}

class NotUrgent extends Urgency {
    sendMessage(message) {
        return "Not Urgent: " + this.notification.send(message);
    }
}

class Urgent extends Urgency {
    sendMessage(message) {
        return "Urgent: " + this.notification.send(message);
    }
}

const email = new Email();
const sms = new SMS();

const urgentEmail = new Urgent(email);
console.log(urgentEmail.sendMessage("API Throttling")); 
// Output: "Urgent: Sending Email: API Throttling"

const notUrgentSMS = new NotUrgent(sms);
console.log(notUrgentSMS.sendMessage("New Blog Post"));
